src/main/java/com/mitchej123/hodgepodge/SimulationDistanceHelper.java

```diff
@@ -0,0 +1,332 @@
+     */
+    public static void preventChunkSimulation(World world, long packedChunkPos, boolean prevent) {
+        if (!FixesConfig.addSimulationDistance) {
+            return;
+        }
+        ISimulationDistanceWorld mixin = (ISimulationDistanceWorld) world;
+        mixin.hodgepodge$preventChunkSimulation(packedChunkPos, prevent);
+    }
+
+    public static int getSimulationDistance() {
+        return TweaksConfig.simulationDistance;
+    }
+
+    public static void setSimulationDistance(int distance) {
+        TweaksConfig.simulationDistance = distance;
+    }
+
+    public SimulationDistanceHelper(World world) {
+        this.world = world;
+        isServer = world instanceof WorldServer;
+    }
+
+    /**
+     * The current world
+     */
+    private final World world;
+
+    /**
+     * True if we have a WorldServer
+     */
+    private final boolean isServer;
+
+    /**
+     * Chunks that should not be ticked
+     */
+    private final Long2ByteOpenHashMap noTickChunks = new Long2ByteOpenHashMap();
+
+    /**
+     * Changes to the chunks that should not be ticked. Needed since we only update the real map once a tick. Reason is
+     * that forced chunks are also cached and updated once a tick, we want both in sync.
+     */
+    private final List<LongBooleanPair> noTickChunksChanges = new ArrayList<>();
+
+    /**
+     * Cache of forced chunks. Profiler says that is faster.
+     */
+    private final LongOpenHashSet forcedChunksMap = new LongOpenHashSet();
+
+    /**
+     * Cache to quickly check if a chunk should be processed or not. Profiler says this is faster.
+     */
+    private final Long2BooleanOpenHashMap cacheShouldProcessTick = new Long2BooleanOpenHashMap();
+
+    /**
+     * Old player positions to see if they moved.
+     */
+    private final Map<EntityPlayer, ChunkCoordIntPair> playerPosOldMap = new HashMap<>();
+
+    /**
+     * All ticks that are currently considered.
+     */
+    private final TreeSet<NextTickListEntry> pendingTickCandidates = new TreeSet<>();
+
+    /**
+     * Map of ticks to chunk, duplicating pendingTickListEntriesTreeSet for quicker lookup.
+     */
+    private final Long2ObjectMap<HashSet<NextTickListEntry>> chunkTickMap = new Long2ObjectOpenHashMap<>();
+
+    /**
+     * TreeSet pendingTickListEntriesTreeSet from WorldServer
+     */
+    private TreeSet<NextTickListEntry> pendingTickListEntriesTreeSet;
+    /**
+     * HashSet pendingTickListEntriesHashSet from WorldServer
+     */
+    private Set<NextTickListEntry> pendingTickListEntriesHashSet;
+    /**
+     * Check if chunk exists in WorldServer
+     */
+    private BiPredicate<Integer, Integer> chunkExists;
+
+    /**
+     * Simulation distance from last tick, to detect if it has changed.
+     */
+    private int simulationDistanceOld;
+
+    public void preventChunkSimulation(long packedChunkPos, boolean prevent) {
+        synchronized (noTickChunksChanges) {
```

Comment:
This should all run singlethreaded, synchronization here is wrong and if you are getting CMEs it won't help because it's most likely caused by mutation during iteration and not multithreading.

Reply:
Thanks, pushed an update.